### ISSUE TITLE
Unable to apply regular expression '(//.*|/\*(.|())*?\*/)' - http://bugs.sun.com/bugdatabase/view_bug.do?bug_id=5050507

### ISSUE THREAD

**Issue by andirdju** [SonarSource/sonar-css#107]
got this error message when running the scan:

```
Unable to lex source code at line : 6666 and column : 0 in file : tests://unittest: 

Unable to apply regular expression '(//.*|/\*(.|())*?\*/)' at line 6666 and column 0, 
because it led to a stack overflow error.

This error may be due to an inefficient use of alternations - see http://bugs.sun.com/bugdatabase/view_bug.do?bug_id=5050507: StackOverflowError
```

**Comment by alban-auzeill** [comment]
I confirm that I can reproduce by parsing a long multi-lines comment, e.g.:
```
/*
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678
012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678
*/
```
The failure is due to this regular expression [CssLexer.java#L40](https://github.com/SonarSource/sonar-css/blob/1.0.0.457/sonar-css-plugin/src/main/java/org/sonar/css/plugin/CssLexer.java#L40) : 
```
String MULTI_LINE_COMMENT = "/\\*(.|" + NEW_LINE + ")*?\\*/";
```

It causes a StackOverflowError containing:
```
Caused by: org.sonar.sslr.channel.ChannelException: Unable to apply regular expression '(//.*|/\*(.|(
|
|))*?\*/)' at line 1 and column 0, because it led to a stack overflow error.
This error may be due to an inefficient use of alternations
 - see http://bugs.sun.com/bugdatabase/view_bug.do?bug_id=5050507
...
Caused by: java.lang.StackOverflowError
	at java.util.regex.Pattern$Branch.match(Pattern.java:4604)
	at java.util.regex.Pattern$GroupHead.match(Pattern.java:4658)
	at java.util.regex.Pattern$LazyLoop.match(Pattern.java:4847)
	at java.util.regex.Pattern$GroupTail.match(Pattern.java:4717)
```

**Comment by Godin** [comment]
@alban-auzeill the problem is exactly as described in mentioned ticket http://bugs.sun.com/bugdatabase/view_bug.do?bug_id=5050507 :

> Avoid alternation whenever possible, alternations are inefficient, causing slow match times as well.

however I would also add that alternations (`|`) generally work fine with possessive quantifiers (i.e. **[that do not backtrack](https://www.regular-expressions.info/possessive.html) and so as a rule of thumb should be preferred, especially when implementing lexers based on regexps** - e.g. `*+`)  and inefficient with other quantifiers (in this our case with reluctant `*?`, in case of mentioned ticket with greedy `*`).

Regex for multi-line comments with possessive quantifier is quite an interesting task, but I'll omit solution as a homework for a curious further reader of this thread, because removal of alternations is simpler - just enable ["dotall mode"](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html#DOTALL):

> In dotall mode, the expression `.` matches any character, including a line terminator. By default this expression does not match line terminators.
> Dotall mode can also be enabled via the embedded flag expression `(?s)`.

In other words:

```
String MULTI_LINE_COMMENT = "(?s)/\\*.*?\\*/";
```

Also I have strong feeling that other regular expressions should be re-reviewed / reworked to use possessive quantifiers and avoid similar problem with other tokens such as big strings.

P.S. to further improve performance and even memory consumption - avoid regexps at all, e.g. by implementing whole lexer as handwritten deterministic finite state machine (trade-off of relative simplicity/understandably and speed of creation/modification of code vs runtime performance)

CC @Evangelink @vilchik-elena FYI

**Comment by Evangelink** [comment]
Same kind of issue for `INLINE_COMMENT`, `DOUBLE_QUOTE_STRING` and `SINGLE_QUOTE_STRING`.
